### PR TITLE
Create view with custom lists to edit

### PR DESCRIPTION
**Pull request by rablador**
When a user decides to edit a custom list they will be sent to a view with all available custom lists. It should look like the view with similar functionality in api access methods. Clicking on one of lists will take the user to an edit screen.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)

### CODE REVIEW

**Review by buggmagnet — CHANGES_REQUESTED** [review]
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: all files reviewed, 4 unresolved discussions (waiting on @rablador)

___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 147 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsmrSYZ0MB_sPiKTuH8:-NsmrSYZ0MB_sPiKTuH9:b-csma37) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L147)):*
> ```Swift
>                 "CUSTOM_LIST_ACTION_SHEET_NEW_LIST",
>                 tableName: "CustomLists",
>                 value: "Add new list",
> ```

Shall we add a `TODO` comment for localisation purposes ?
___
*[`ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift` line 25 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsmsKQ-0pgIhH7Q4Xq1:-NsmsKQ-0pgIhH7Q4Xq2:b-6hkxth) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift#L25)):*
> ```Swift
> 
>     func saveCustomList(viewModel: CustomListViewModel) throws {
>         try _ = repository.save(list: viewModel.customList)
> ```

Did you mean `try repository.save(list: viewModel.customList)` ?
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 97 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsnGJUP1Gr_R4nc83v3:-NsnGJUP1Gr_R4nc83v4:b-n8rpnb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L97)):*
> ```Swift
>             "LIST_CUSTOM_LIST_NAVIGATION_TITLE",
>             tableName: "CustomList",
>             value: "Edit custom list",
> ```

Let's add / reuse localized strings here as well
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 101 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsnHN5V5rj9Tx4Fb3HN:-NsnHN5V5rj9Tx4Fb3HO:by6x9mo) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L101)):*
> ```Swift
>         )
> 
>         navigationItem.rightBarButtonItem = UIBarButtonItem(
> ```

If I'm not mistaken, `rightBarButtonItem` has a `strong` reference to the `UIBarButtonItem` which means by extension,  a strong reference to the closure passed to `UIAction`'s `handler` and therefore, to self.

Long story short, I think we want to do the following instead
```swift
primaryAction: UIAction(handler: {[weak self] _ in
    self?.didFinish?()
})
```

**Review by mojganii — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: all files reviewed, 8 unresolved discussions (waiting on @buggmagnet and @rablador)

___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 131 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsqzX6aAsVjZzIpSq7N:-NsqzX6aAsVjZzIpSq7O:bq6w1ls) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L131)):*
> ```Swift
>     }
> 
>     private func showCustomListActionSheet() {
> ```

nit: do really we need to make a function for this which is called once? IMO this can be placed into `didRequestRouteToCustomLists`
___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 145 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsqyrzFDxIx8Gudp-SZ:-NsqyrzFDxIx8Gudp-S_:b-60aco) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L145)):*
> ```Swift
>         actionSheet.addAction(UIAlertAction(
>             title: NSLocalizedString(
>                 "CUSTOM_LIST_ACTION_SHEET_NEW_LIST",
> ```

it seems the sentence is not desprictive. 



_Code snippet:_
```Swift
  actionSheet.addAction(UIAlertAction(
            title: NSLocalizedString(
                "CUSTOM_LIST_ACTION_SHEET_ADD_NEW_LIST",
                tableName: "CustomLists",
                value: "Add new list",
                comment: ""
            ),
            style: .default,
            handler: { _ in
                self.showAddCustomList()
            }
        ))
```
___
*[`ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift` line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsr0e7r7yZwYaYdlQrj:-Nsr0e7r7yZwYaYdlQrk:by676ym) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift#L12)):*
> ```Swift
> 
> protocol CustomListInteractorProtocol {
>     func fetchAllCustomLists() -> [CustomList]
> ```

`CustomList` keywords in function names are redundant .
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 101 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsnHN5V5rj9Tx4Fb3HN:-Nsr2rFiDcRWBtid8LXm:b32v57) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L101)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

If I'm not mistaken, `rightBarButtonItem` has a `strong` reference to the `UIBarButtonItem` which means by extension,  a strong reference to the closure passed to `UIAction`'s `handler` and therefore, to self.

Long story short, I think we want to do the following instead
```swift
primaryAction: UIAction(handler: {[weak self] _ in
    self?.didFinish?()
})
```
</blockquote></details>

+1
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 145 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsr2y11AW2oZ9uOL3-t:-Nsr2y11AW2oZ9uOL3-u:b-c7zqse) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L145)):*
> ```Swift
> extension ListCustomListViewController: UITableViewDelegate {
>     func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
>         UIMetrics.SettingsCell.apiAccessCellHeight
> ```

it has noting to do with `api access`.let's table view use `UITableView.automaticDimension` or set hardcoded value.

**Review by mojganii — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: all files reviewed, 8 unresolved discussions (waiting on @buggmagnet and @rablador)

___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 145 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsr2y11AW2oZ9uOL3-t:-Nsr4OXJDnnrF33StJEw:bu25f3z) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L145)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

it has noting to do with `api access`.let's table view use `UITableView.automaticDimension` or set hardcoded value.
</blockquote></details>

moreover, if all cells have the same height I would suggest to do that through `tableView.rowHeight = X` instead of recalculating them every time by `tableview`

**Review by rablador — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: 5 of 24 files reviewed, 7 unresolved discussions (waiting on @buggmagnet and @mojganii)

___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 131 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsqzX6aAsVjZzIpSq7N:-NsrjfM3ECav1u8GpQWN:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L131)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

nit: do really we need to make a function for this which is called once? IMO this can be placed into `didRequestRouteToCustomLists`
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 145 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsqyrzFDxIx8Gudp-SZ:-Nsrj_21BAbCabx7sDE_:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L145)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

it seems the sentence is not desprictive.
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 147 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsmrSYZ0MB_sPiKTuH8:-NsrjiG3-ynmaLRLyxmT:b-tf9qk3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L147)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Shall we add a `TODO` comment for localisation purposes ?
</blockquote></details>

Not sure I follow.
___
*[`ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift` line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsr0e7r7yZwYaYdlQrj:-Nsrk-qf84xNvRI7RqRE:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift#L12)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

`CustomList` keywords in function names are redundant .
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift` line 25 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsmsKQ-0pgIhH7Q4Xq1:-NsrkleL4QNHSsUpVKyF:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/CustomListInteractor.swift#L25)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Did you mean `try repository.save(list: viewModel.customList)` ?
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 97 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsnGJUP1Gr_R4nc83v3:-Nsrl4B7EECDyWoK1l3c:b-h1q2xy) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L97)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Let's add / reuse localized strings here as well
</blockquote></details>

Resuse from where?
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 101 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsnHN5V5rj9Tx4Fb3HN:-NsrlUyJAoSFQNMG1H6o:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L101)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift` line 145 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsr2y11AW2oZ9uOL3-t:-NsrmN7AAUeYfolPQ1lQ:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListViewController.swift#L145)):*
<details><summary><i>Previously, mojganii wrote…</i></summary><blockquote>

moreover, if all cells have the same height I would suggest to do that through `tableView.rowHeight = X` instead of recalculating them every time by `tableview`
</blockquote></details>

Done.

**Review by rablador — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: 5 of 24 files reviewed, 8 unresolved discussions (waiting on @buggmagnet and @mojganii)

___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListCoordinator.swift` line 73 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NssD-vN8YaEfmZd9wxN:-NssD-vODQvDVX5fB6b8:b-vq35am) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/c6c49c798c3266d99b1c41da7a7a0900b3752d1b/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListCoordinator.swift#L73)):*
> ```Swift
>     }
> 
>     private func updateRelayConstraints(for action: EditCustomListCoordinator.FinishAction, in list: CustomList) {
> ```

Makes sure we update the relay constraints / relay selector whenever we make changes to a list we're currently using.

**Review by rablador — COMMENTED** [review]
> Dismissed @buggmagnet from a discussion.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: 20 of 25 files reviewed, 8 unresolved discussions (waiting on @buggmagnet and @mojganii)

___
*[`ios/MullvadTypes/RelayConstraints.swift` line 32 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsw0ylODihZnYHd5yYK:-Nsw4RYR5XIsc3zhvk2i:bjwqan7) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/58b1b136371f0693eade6ef75dd91602082d9f79/ios/MullvadTypes/RelayConstraints.swift#L32)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

As it stands right now, if the user updates their application whilst being connected, it will force them into blocked mode because their currently selected relay constraints will not match anymore. 

Can we prevent this from happening ?
</blockquote></details>

I'll do a migration for it.
___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 147 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NsmrSYZ0MB_sPiKTuH8:-NswUSl-4LQIctbAMeIT:byent7v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/68903ea2bcfd530f79b07c57ac537aeb230a43c8/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L147)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Nevermind, I think I misread the original code
</blockquote></details>

Ok, no change then.
___
*[`ios/MullvadVPN/Coordinators/LocationCoordinator.swift` line 45 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsw1HaQ17Qf5-jxR8yA:-NswUR6fB54h4Ha94u_g:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/58b1b136371f0693eade6ef75dd91602082d9f79/ios/MullvadVPN/Coordinators/LocationCoordinator.swift#L45)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

We probably want to remove this
</blockquote></details>

Done.
___
*[`ios/MullvadVPN/Coordinators/CustomLists/ListCustomListCoordinator.swift` line 48 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsw1hX3Ekvi5PtHUKwp:-NswUZvP1ghU-Ir40oPX:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/58b1b136371f0693eade6ef75dd91602082d9f79/ios/MullvadVPN/Coordinators/CustomLists/ListCustomListCoordinator.swift#L48)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Let's remove this too
</blockquote></details>

Done.
___
*[`ios/MullvadVPN.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved` line 22 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-Nsw0fxe1r5RbS2XpAx0:-NswUbzV55bjZvPuct_k:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/58b1b136371f0693eade6ef75dd91602082d9f79/ios/MullvadVPN.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved#L22)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

This comes from Xcode 15.3 which we do not use yet in the CI, please revert this file with the following command
`git checkout main -- MullvadVPN.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/`
</blockquote></details>

Done.

**Review by rablador — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909)* status: 20 of 25 files reviewed, 9 unresolved discussions (waiting on @buggmagnet and @mojganii)

___
*[`ios/MullvadTypes/RelayLocation.swift` line 134 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-NswVEVsAsDgNqIRC4ge:-NswVEVsAsDgNqIRC4gf:bqt64ge) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/0226a1485fda41bbaba4e94c088c6bdc9287a68e/ios/MullvadTypes/RelayLocation.swift#L134)):*
> ```Swift
> }
> 
> /// - Warning: Deprecated, use UserSelectedRelays instead.
> ```

Opting not to use `@available(*, deprecated)` here since it will forever give us a deprecation warning when migrating from `RelayLocations` to `UserSelectedLocations` in `RelayConstraints`.

**Review by buggmagnet — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-:-Nswqtcb6dQGs_z1E22e:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @mojganii)

**Review by buggmagnet — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-:-Nsx6kk81xBWn-pFclCx:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @mojganii)

**Review by mojganii — CHANGES_REQUESTED** [review]
when I delete a custom list which is selected as an active RelayConstarints the app doesn't go to block state.

> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-:-NsxKIMS8dXvzZEuP65R:b-ma9s0i)* status: 24 of 26 files reviewed, 5 unresolved discussions (waiting on @buggmagnet)

**Review by buggmagnet — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-:-Nt0-i7R-VvjCCnyNz0_:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @mojganii)

**Review by mojganii — APPROVED** [review]
In the real device it goes on blocked state.

> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5909#-:-Nt0CvqkDO3Ut-KkQ0JY:b-63e08g)* status: all files reviewed, 4 unresolved discussions
